renpy.el

```diff
@@ -1528,6 +1533,185 @@ Uses `renpy-beginning-of-block', `renpy-end-of-block'."
+      (setq whitespace (not (= (skip-syntax-backward " ") 0)))
+      (skip-syntax-backward "w_")
+      (setq candidate (thing-at-point 'symbol t))
+      (if (member candidate renpy--completion-keywords)
+	  (setq keyword candidate)
+	(setq prev-pos (point))))
+    (goto-char prev-pos)
+    (when whitespace keyword)))
+
+(defun renpy--completion-context ()
+  "Return the completion category keyword symbol at point."
+  ;; TODO: More specific contexts: show text, show expression, ...
+  (when (renpy--completion-context-check)
+    (save-excursion
+      (let ((prev (renpy--skip-to-keyword-backward)))
+	(cond
+	 ;; call/jump <point>
+	 ((member prev '("call" "jump")) :label)
+	 ;; show/scene/hide <point>
+	 ((member prev '("show" "scene" "hide")) :image)
+	 ;; at <point>
+	 ((equal prev "at") :transform))))))
+
+(defconst renpy--label-definition-re
+  (renpy-rx label-keyword (1+ space) (group label-name))
+  "Regexp for looking up label definitions.")
+
+(defun renpy--collect-labels ()
+  "Return all unique label names in the current buffer."
+  (let (labels)
+    (save-excursion
+      (goto-char (point-min))
+      (while (re-search-forward renpy--label-definition-re nil t)
+        (push (match-string-no-properties 1) labels)))
+    labels))
+
+(defconst renpy--image-definition-re
+  (renpy-rx image-keyword (1+ space)
+	    (group image-name		; tag
+		   (0+ space name))	; attributes
+	    (0+ space) (or "=" ":"))
+  "Regexp for looking up image definitions.")
+
+(defun renpy--collect-images ()
+  "Return all unique image names in the current buffer."
+  (let (images)
+    (save-excursion
+      (goto-char (point-min))
+      (while (re-search-forward renpy--image-definition-re nil t)
+        (push (match-string-no-properties 1) images)))
+    images))
+
+(defconst renpy--transform-definition-re
+  (renpy-rx transform-keyword (1+ space) (group name))
+  "Regexp for looking up transform definitions.")
+
+(defun renpy--collect-transforms ()
+  "Return all unique transform names in the current buffer."
+  (let (transforms)
+    (save-excursion
+      (goto-char (point-min))
+      (while (re-search-forward renpy--transform-definition-re nil t)
+        (push (match-string-no-properties 1) transforms)))
+    transforms))
+
+(defun renpy-completion-at-point ()
+  "Provide completion data for the symbol at point in Ren'Py buffers."
+  (let (candidates beg end)
+    (pcase (renpy--completion-context)
+      (:label
+       (setq end (point)
+	     beg (save-excursion
+		   (skip-syntax-backward "w_")
+		   (point))
+	     candidates (renpy--collect-labels)))
+      (:image
+       (setq end (point)
+	     beg (save-excursion
+		   (renpy--skip-to-keyword-backward)
+		   (point))
+	     candidates (renpy--collect-images)))
+      (:transform
+       (setq end (point)
+	     beg (save-excursion
+		   (skip-syntax-backward "w_")
+		   (point))
+	     candidates (renpy--collect-transforms))))
+    (when candidates
```

Comment:
Same here, for preferring `and` over `when`, when it is the return value that you are interested in rather than a side-effect.